README.md
# alda-server-clj

A Clojure implementation of an [Alda](https://github.com/alda-lang/alda) server.

## Roles

- Responds to requests from an Alda client.
- Manages Alda worker processes.

## Development

### Overview

The Alda server handles a variety of things, including parsing Alda code into executable Clojure code, executing the code to create or modify a score, modeling a musical score as a Clojure map of data, and using the map of data as input to interpret the score (generating sound).

The Alda server spawns external *worker* processes and delegates most of the aforementioned work to those processes. The processes communicate back and forth via [ZeroMQ](https://zeromq.org) sockets. For more information about the use of ZeroMQ in Alda processes, see our [ZeroMQ Architecture](https://github.com/alda-lang/alda/blob/master/doc/zeromq-architecture.md) document.

### Prerequisites

Development on the Alda server and worker processes requires that you have the [Boot](http://boot-clj.com) build tool installed. This allows you to run the `test` and `dev` tasks to run tests and run the server/worker locally for development.

### `boot test` task

To run the integration test suite, run `boot test`. This will test that the server and worker processes are communicating with each other correctly.

### `boot dev` task

The `dev` task allows you to run the Alda server and worker processes locally, including any changes you have made to the codebase.

Note that the Alda client only communicates with the server, never directly to the worker. This means that to test changes, you will often need to run both the server and the worker at the same time, and make sure the worker is using the correct port number to talk to the server backend. This setup is described in detail below.

#### Alda Server

To run an Alda server locally:

    $ boot dev -a server --port 27714 --alda-fingerprint

The `--port` and `--alda-fingerprint` arguments are strictly optional, but including them will ensure that the Alda client recognizes your development server as an Alda server and includes it in the output of `alda list`.

#### Alda Worker

To run an Alda worker process locally:

    $ boot dev -a worker --port 12345

The `--port` argument needs to be the backend port on which a server is managing its workers. You can see this in the output of the server when it starts up:

    $ boot dev -a server --port 27714 --alda-fingerprint
    ...
    16-Sep-04 15:21:49 skeggox.local INFO [alda.server] - Binding frontend socket on port 27714...
    16-Sep-04 15:21:49 skeggox.local INFO [alda.server] - Binding backend socket on port 60610...
    16-Sep-04 15:21:49 skeggox.local INFO [alda.server] - Spawning 4 workers...

    # in another terminal
    $ boot dev -a worker --port 60610 --alda-fingerprint
    ...
    16-Sep-04 15:23:15 skeggox.local INFO [alda.worker] - Logging errors to /Users/dave/.alda/logs/error.log
    Sep 04, 2016 3:23:15 PM com.jsyn.engine.SynthesisEngine start
    INFO: Pure Java JSyn from www.softsynth.com, rate = 44100, RT, V16.7.3 (build 457, 2014-12-25)

#### Tricking the Supervisor

The server has a "supervisor" routine that it does every so often to make sure that it still has the correct number of workers. If you start your own worker process in addition to the workers that the server spawned, then the server will have one more worker than it needs and will "lay off" one worker. The worker it lays off could be your debug worker, which could get in the way of what you're trying to do.

To prevent your debug worker process from being laid off, set the environment variable `ALDA_DISABLE_SUPERVISOR` when starting the server. You can also set the number of workers spawned by the server to 0 in order to ensure that your worker will receive all of the requests.

You may also want to set `ALDA_DEBUG_MODE` when starting the worker in order to see debug-level logs printed to the console, instead of only error logs logged to `~/.alda/logs/error.log` (instead of printed).

    $ ALDA_DISABLE_SUPERVISOR=yes boot dev -a server \
                                           --port 27714 \
                                           --workers 0 \
                                           --alda-fingerprint
    ...
    16-Sep-04 15:32:59 skeggox.local INFO [alda.server] - Binding frontend socket on port 27714...
    16-Sep-04 15:32:59 skeggox.local INFO [alda.server] - Binding backend socket on port 60830...
    16-Sep-04 15:32:59 skeggox.local INFO [alda.server] - Spawning 0 workers...

    # in a separate terminal
    $ ALDA_DEBUG_MODE=yes boot dev -a worker --port 60830 --alda-fingerprint
    ...
    16-Sep-04 15:34:55 skeggox.local INFO [alda.worker] - Loading Alda environment...
    Sep 04, 2016 3:34:55 PM com.jsyn.engine.SynthesisEngine start
    INFO: Pure Java JSyn from www.softsynth.com, rate = 44100, RT, V16.7.3 (build 457, 2014-12-25)
    16-Sep-04 15:34:58 skeggox.local INFO [alda.worker] - Worker reporting for duty!
    16-Sep-04 15:34:58 skeggox.local INFO [alda.worker] - Connecting to socket on port 60864...
    16-Sep-04 15:34:58 skeggox.local INFO [alda.worker] - Sending READY signal.
    16-Sep-04 15:34:58 skeggox.local DEBUG [alda.worker] - Got HEARTBEAT from server.
    16-Sep-04 15:34:59 skeggox.local DEBUG [alda.worker] - Got HEARTBEAT from server.
    16-Sep-04 15:35:00 skeggox.local DEBUG [alda.worker] - Got HEARTBEAT from server.


## License

Copyright © 2012-2019 Dave Yarwood et al

Distributed under the Eclipse Public License version 1.0.


CHANGELOG.md
# CHANGELOG

## 0.5.0 (2019-04-13)

* Renamed (not yet used in an Alda release) `export-status` command to
  `job-status`.

  I've convinced myself at this point that it's equivalent to the `play-status`
  command, so `play-status` is just a supported synonym at this point. In the
  next release of Alda, the client will send a `job-status` request for both
  `play` and `export` jobs.

## 0.4.4 (2019-04-11)

* Added support for MIDI `export` and `export-status` commands.

## 0.4.3 (2019-01-28)

* Refactored to remove dependency on the `alda.now` namespace, which is being
  removed from `alda/sound-engine-clj`. We can achieve the same functionality in
  `alda/server-clj` by using `alda.sound` directly.

## 0.4.2 (2019-01-19)

* Add support for an `instruments` command, which returns a list of available
  instruments.

## 0.4.1 (2018-10-28)

* Updates to support alda/sound-engine-clj 1.0.0, which now uses a Java MIDI
  Sequencer instead of JSyn to do event scheduling:

  * Removed call to `alda.sound/start-synthesis-engine!`, which has been
    removed.

  * Added a call to `midi/open-midi-sequencer!` when starting a worker process.

## 0.4.0 (2018-06-22)

* Removed a stale reference to scheduled functions, [for compatibility with
  alda/core 0.4.0](https://github.com/alda-lang/alda-core/pull/65).

## 0.3.3 (2017-10-07)

* Added support for an "events" option to the "parse" command.

## 0.3.2 (2017-10-05)

* Added an `alda.server/*disable-supervisor*` dynamic var that, when bound to a
  truthy value, has the same effect as the `ALDA_DISABLE_SUPERVISOR` environment
  variable.

  I needed this in order to write a convenient development task for the
  alda/core project. The Boot process was already running a JVM, so I couldn't
  really set `ALDA_DISABLE_SUPERVISOR` after the fact.

* Increased the number of worker lives from 10 to 30. This effectively means
  that a worker will wait until it's been 30 seconds, not 10 seconds, since it
  received its last heartbeat from the server before giving up and killing
  itself. This should help with some timing issues I was seeing where the server
  took a little longer than 10 seconds to send a worker its first heartbeat. 30
  seconds ought to be a safer timeframe.

## 0.3.1 (2017-06-11)

* Added support for the `stop-playback` command. When a server gets that
  command, it sends a `STOP` signal to all known workers. Each worker then stops
  playback immediately.

## 0.3.0 (2017-05-27)

* Updates for compatibility with alda/core 0.2.0, which has a brand new parser
  rewritten from the ground up.

  One thing about the new parser is that it doesn't generate alda.lisp code
  anymore, so that feature has been removed from the worker.

* Bugfix: scores that included scheduled Clojure function calls were breaking
  JSON serialization. In the long term, I need to think about how these can be
  serialized, but in the short term, we don't really need them in the score
  JSON, so as a hack, we just aren't including them in the score the server
  returns.

## 0.2.0 (2017-05-25)

* Updated ezzmq dependency to 0.5.3. This updates the transitive JeroMQ
  dependency from 0.3.6 to 0.4.0.

* Partially solves [#8](https://github.com/alda-lang/alda-server-clj/issues/8)
  by implementing a job system. Requests from the client must include a `jobId`
  (a UUID generated by the client), which allows the worker to track jobs
  separately.

  A worker should only be processing one job at a time, but it's possible for
  requests to arrive out of sync, so we want to make sure that a `play` request
  followed by a `play-status` request will receive status for the correct score.

  There is still some work to be done here. I'd like to move the job cache into
  the server, so that the server can deduplicate multiple requests to play the
  same score (i.e. same `jobId`). The worker will do that, as of this release,
  but it doesn't really matter because multiple requests to play the same score
  will be handled by multiple workers, so each worker will still play the score,
  thinking the job hasn't been done yet. The client is currently only sending
  one request to play a score, so multiple play requests aren't sent, however it
  would be nice to be able to retry on the client side. This will be a work in
  progress.

* Bugfix: don't reuse the same MIDI synthesizer instance (which now has a dirty
  state) after playing a score.

## 0.1.7 (2017-05-19)

* Each worker process now has a lifespan of 15-20 minutes. Once its lifespan is
  up, it finishes doing whatever work it might be doing and then shuts down so
  that the server can replace it with a fresh worker.

  This ensures that available workers are always fresh (spawned within the last
  20 minutes), which helps us avoid audio errors like delayed MIDI audio.

  Fixes issue [#5](https://github.com/alda-lang/alda-server-clj/issues/5).

## 0.1.6 (2017-05-18)

* Improved the reliability of including the score map in `play-status` responses.

  Prior to this version, we were storing the state of a worker in three separate
  (isolated) atoms: `current-status`, `current-score`, and `current-error`.
  Because we were storing state this way, it was possible for one state to be
  updated independently of another, when we actually wanted to update them at
  the same time as part of a transaction. For example, the status could be
  updated to "playing," and then the worker could answer a request for status
  before it updated the score to the score it responds with, so the status and
  score could be out of sync.

  Clojure provides the `ref` type to solve problems like this; they allow you to
  update multiple values transactionally. As of this version, we are using refs
  instead of atoms, which should make the `play-status` responses more
  consistent.

## 0.1.5 (2017-05-17)

* Include the score map (as a JSON string) in response to `play-status`
  requests, if the score has been parsed at that point.


test/alda/worker_test.clj
(ns alda.worker-test
  (:require [clojure.test  :refer :all]
            [cheshire.core :as    json]
            [ezzmq.core    :as    zmq]
            [alda.worker   :as    worker]
            [alda.zmq-util :refer (find-open-port)]
            [alda.version  :refer (-version-)])
  (:import [java.nio.file Files Paths]
           [java.nio.file.attribute FileAttribute]))

(def ^:dynamic *port*   nil)
(def ^:dynamic *socket* nil)

(defn- init!
  [var val]
  (alter-var-root var (constantly val)))

(defn- generate-job-id
  []
  (letfn [(rand-char [] (char (rand-nth (range (int \a) (inc (int \z))))))]
    (apply str (repeatedly 20 rand-char))))

(defn response-for-msg
  [msg]
  (zmq/send-msg *socket* msg)
  (-> (zmq/receive-msg *socket* :stringify true)
      second
      (json/parse-string true)))

(defn response-for
  [{:keys [command] :as req}]
  (let [msg [(json/generate-string req) command]]
    (response-for-msg msg)))

(defn complete-response-for
  [{:keys [command] :as req}]
  (let [msg [(json/generate-string req) command]]
    (zmq/send-msg *socket* msg))
  (zmq/receive-msg *socket* :stringify true))

(defn response-for-msg
  [msg]
  (zmq/send-msg *socket* msg)
  (loop [res (zmq/receive-msg *socket* :stringify true)]
    (if (#{"READY" "AVAILABLE" "BUSY"} (first res))
      (recur (zmq/receive-msg *socket* :stringify true))
      res)))

(defn response-for
  [{:keys [command] :as req}]
  (let [msg [(.getBytes "fakeclientaddress")
             (json/generate-string req)
             command]]
    (response-for-msg msg)))

(use-fixtures :once
  (fn [run-tests]
    (zmq/with-new-context
      (init! #'*port*   (find-open-port))
      (init! #'*socket* (zmq/socket
                          :dealer
                          {:bind (format "tcp://*:%s" *port*)}))
      (zmq/worker-thread {:on-interrupt #(println "Worker interrupted!")}
        (Thread/sleep 1000) ; to make sure we don't miss the READY signal
        (binding [worker/*no-system-exit* true]
          (worker/start-worker! *port* true)))
      (run-tests))))

(deftest worker-tests
  (testing "a worker process"
    (testing "should send a READY signal when it starts up"
      (let [[msg] (zmq/receive-msg *socket* :stringify true)]
        (is (= "READY" msg))))
    (testing "should send AVAILABLE heartbeats while waiting for work"
      (dotimes [_ 5]
        (let [[msg] (zmq/receive-msg *socket* :stringify true)]
          (is (= "AVAILABLE" msg)))))
    ;; NOTE: the expected responses are tested in alda.server-test
    (testing "should successfully respond to"
      (testing "a 'parse' command"
        (let [req {:command "parse"
                   :body "piano: c8 d e f g2"}
              [_ _ json :as res] (response-for req)
              {:keys [success body]} (json/parse-string json true)]
          (is success body)))
      (let [job-id (generate-job-id)]
        (testing "a 'play' command"
          (let [req {:command "play"
                     :body "piano: (vol 0) c2"
                     :options {:jobId job-id}}
                [_ _ json] (response-for req)
                {:keys [success body]} (json/parse-string json true)]
            (is success body)))
        ;; TIMING: give the worker time to start processing the job
        (Thread/sleep 1000)
        (testing "a 'play-status' command"
          (let [req {:command "play-status" :options {:jobId job-id}}
                [_ _ json] (response-for req)
                {:keys [success body] :as res} (json/parse-string json true)]
            (is success body))))
      (let [job-id (generate-job-id)]
        (testing "an 'export' command"
          (let [tmp-dir  (Files/createTempDirectory
                           "alda-worker-test"
                           (into-array FileAttribute []))
                filename (-> tmp-dir
                             .toString
                             (Paths/get (into-array String [job-id]))
                             .toAbsolutePath
                             (str ".mid"))
                req {:command "export"
                     :body    "piano: (vol 0) c2"
                     :options {:jobId job-id, :filename filename}}
                [_ _ json]             (response-for req)
                {:keys [success body]} (json/parse-string json true)]
            (is success body)))
        ;; TIMING: give the worker time to start processing the job
        (Thread/sleep 1000)
        (testing "a 'job-status' command"
          (let [req {:command "job-status" :options {:jobId job-id}}
                [_ _ json] (response-for req)
                {:keys [success body] :as res} (json/parse-string json true)]
            (is success body)))))
    (testing "should accept signals from the server"
      (doseq [signal ["HEARTBEAT" "STOP" "KILL"]]
        (zmq/send-msg *socket* signal)))))


src/alda/queue.clj
(ns alda.queue)

(defn queue
  "A janky custom data structure that acts like a queue and is a ref.

   It's really just a vector wrapped in a ref.

   Items are popped from the left and pushed onto the right."
  ([]
   (ref []))
  ([init]
   (ref (into [] init))))

(defn push-queue
  [q x]
  (alter q #(conj % x)))

(defn pop-queue
  "Pops off the first item in the queue and returns it."
  [q]
  (let [x (first @q)]
    (alter q #(vec (drop 1 %)))
    x))

(defn reverse-pop-queue
  "Pops off the LAST item in the queue and returns it."
  [q]
  (let [x (last @q)]
    (alter q #(vec (drop-last 1 %)))
    x))

(defn check-queue
  "Returns true if there is at least one item in the queue that satisfies the
   predicate."
  [q pred]
  (boolean (some pred @q)))

(defn re-queue
  "Finds all items in the queue that satisfy the predicate, and re-queues them
   onto the end of the queue.

   When a second function `f` is provided, it is called on each re-queued
   element. This can be used e.g. to update the timestamps of queued items."
  [q pred & [f]]
  (alter q #(let [yes (for [x % :when (pred x)]
                        ((or f identity) x))
                  no  (filter (complement pred) %)]
              (vec (concat no yes)))))

(defn remove-from-queue
  "Removes all items from the queue that satisfy the predicate."
  [q pred]
  (alter q #(vec (filter (complement pred) %))))



src/alda/worker.clj
(ns alda.worker
  (:require [alda.parser                :refer (parse-input)]
            [alda.lisp.instruments.midi :refer (instruments)]
            [alda.lisp.score            :as    score]
            [alda.sound                 :as    sound]
            [alda.sound.midi            :as    midi]
            [alda.util                  :as    util]
            [alda.version               :refer (-version-)]
            [cheshire.core              :as    json]
            [clojure.core.cache         :as    cache]
            [clojure.set                :as    set]
            [taoensso.timbre            :as    log]
            [ezzmq.core                 :as    zmq]))

(def ^:dynamic *no-system-exit* false)

(defn exit!
  [exit-code]
  (when-not *no-system-exit* (System/exit exit-code)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn start-alda-environment!
  []
  (midi/open-midi-synth!)
  (midi/open-midi-sequencer!)
  (log/debug "Requiring alda.lisp...")
  (require '[alda.lisp :refer :all]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn- success-response
  [body]
  {:success true
   :body    (if (or (map? body) (sequential? body))
              (json/generate-string body)
              body)})

(defn- error-response
  [e]
  {:success false
   :body    (if (string? e)
              e
              (.getMessage e))})

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; Keep information about each job in memory for 2 hours.
(def ^:const JOB_CACHE_TTL (* 2 60 60 1000))

(defrecord Job [status score error stop!])

(def job-cache (atom (cache/ttl-cache-factory {} :ttl JOB_CACHE_TTL)))

(defn update-job! [id job]
  "Upserts a Job record into the cache.
   If a job with that ID is already in the cache, updates it.
   Otherwise, adds it to the cache."
  (swap! job-cache (fn [c]
                     (if (cache/has? c id)
                       (-> (cache/hit c id)
                           (assoc id job))
                       (cache/miss c id job)))))

(defn update-job-status! [id status]
  "Updates the `status` of a job, preserving the existing values of `score` and
   `error`.

   If the job doesn't exist yet, adds a new job with the provided `id` and
   `status`."
  (swap! job-cache (fn [c]
                     (if (cache/has? c id)
                       (-> (cache/hit c id)
                           (update id assoc :status status))
                       (cache/miss c id (Job. status nil nil nil))))))

(defn pending?
  [{:keys [status] :as job}]
  (not (#{:success :error :playing :exporting} status)))

(defn available?
  []
  (not-any? #(#{:parsing :playing :exporting} (:status %)) (vals @job-cache)))

(defn run-play-job!
  [code {:keys [history from to jobId]}]
  (try
    (log/debugf "Starting job %s..." jobId)
    (update-job! jobId (Job. :parsing nil nil nil))
    (let [events  (do
                    (log/debug "Parsing body...")
                    (parse-input code :output :events-or-error))
          history (when history
                    (log/debug "Parsing history...")
                    (parse-input history))
          score   (do
                    (log/debug "Constructing score...")
                    (apply score/continue
                           (or history (score/new-score))
                           events))

          {:keys [score stop! wait]}
          (do
            (log/debug "Playing score...")
            (sound/with-play-opts {:from     from
                                   :to       to
                                   :async?   true
                                   :one-off? false}
              (sound/play!
                score
                (when (seq (:events history))
                  (set/difference (:events score) (:events history))))))]
      (update-job! jobId (Job. :playing score nil stop!))
      (wait)
      (log/debug "Done playing score.")
      (update-job-status! jobId :success))
    (catch Throwable e
      (log/error e e)
      (update-job! jobId (Job. :error nil e nil)))))

(defn run-export-job!
  [code {:keys [jobId filename]}]
  (try
    (log/debugf "Starting job %s..." jobId)
    (update-job! jobId (Job. :parsing nil nil nil))
    (log/debug "Parsing score...")
    (let [score (parse-input code)]
      (update-job! jobId (Job. :exporting score nil nil))
      (log/debugf "Exporting score to %s..." filename)
      (sound/export! score filename)
      (log/debug "Done exporting score.")
      (update-job-status! jobId :success))
    (catch Throwable e
      (log/error e e)
      (update-job! jobId (Job. :error nil e nil)))))

(defn stop-playback!
  "Stops playback for all jobs where that is possible (i.e. ones that have a
   `:stop!` function)."
  []
  (doseq [[id {:keys [stop!] :as job}] @job-cache
          :when stop!]
    (stop!)
    (update-job! id (assoc job :status :success))))

(defn handle-code-play
  [code {:keys [jobId] :as options}]
  (-> (cond
        (empty? jobId)
        (error-response "Request missing a `jobId` option.")

        (get @job-cache jobId)
        (success-response "Already playing that score.")

        :else
        (do
          (future (run-play-job! code options))
          (success-response "Request received.")))
      (assoc :jobId jobId)))

(defn handle-code-export
  [code {:keys [jobId filename] :as options}]
  (-> (cond
        (empty? jobId)
        (error-response "Request missing a `jobId` option.")

        (get @job-cache jobId)
        (success-response "Already exporting that score.")

        :else
        (do
          (future (run-export-job! code options))
          (success-response "Request received.")))
      (assoc :jobId jobId)))

(defn handle-code-parse
  [code {:keys [output] :as options}]
  (try
    (require '[alda.lisp :refer :all])
    (let [output (case output
                   ("data" nil) :score
                   "events"     :events-or-error)]
      (success-response (parse-input code :output output)))
    (catch Throwable e
      (log/error e e)
      (error-response e))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defmulti process :command)

(defmethod process :default
  [{:keys [command]}]
  (error-response (format "Unrecognized command: %s" command)))

(defmethod process nil
  [_]
  (error-response "Missing command"))

(defmethod process "parse"
  [{:keys [body options]}]
  (handle-code-parse body options))

(defmethod process "ping"
  [_]
  (success-response "OK"))

(defmethod process "play"
  [{:keys [body options]}]
  (handle-code-play body options))

(defmethod process "export"
  [{:keys [body options]}]
  (handle-code-export body options))

(defn- sanitize-score-for-json
  "We need to convert the score (a Clojure map) into a JSON string to send as a
   response, but certain values in the map are not serializable as JSON."
  [score]
  (-> score
      ;; remove the audio-context (an atom containing information specific to
      ;; playing the score)
      (dissoc :audio-context)))

(defn job-status
  [{:keys [options]}]
  (let [job-id (get options :jobId)
        {:keys [status score error] :as job} (get @job-cache job-id)]
    (-> (cond
          (empty? job-id)
          (error-response "Request missing a `jobId` option.")

          (nil? status)
          (error-response "Job not found.")

          (= :error status)
          (error-response error)

          :else
          (-> (success-response (name status))
              (assoc :score (sanitize-score-for-json score))
              (assoc :pending (pending? job))))
        (assoc :jobId job-id))))

(defmethod process "play-status"
  [msg]
  (job-status msg))

(defmethod process "job-status"
  [msg]
  (job-status msg))

(defmethod process "version"
  [_]
  (success-response -version-))

(defmethod process "instruments"
  [_]
  (-> (success-response "OK")
      (assoc :instruments (map first instruments))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(def ^:const MIN-LIFESPAN       (* 1000 60 15)) ; 15 minutes
(def ^:const MAX-LIFESPAN       (* 1000 60 20)) ; 20 minutes
(def ^:const HEARTBEAT-INTERVAL 1000)  ; 1 second
(def ^:const SUSPENDED-INTERVAL 10000) ; 10 seconds
(def ^:const MAX-LIVES          30)

(def lives (atom MAX-LIVES))

(defn start-worker!
  [port & [verbose?]]
  (util/set-log-level! :debug)
  (when-not (or (System/getenv "ALDA_DEBUG") verbose?)
    (let [log-path (util/alda-home-path "logs" "error.log")]
      (log/infof "Logging errors to %s" log-path)
      (util/set-log-level! :error)
      (util/rolling-log! log-path)))
  (start-alda-environment!)
  (log/info "Worker reporting for duty!")
  (log/infof "Connecting to socket on port %d..." port)
  (zmq/with-new-context
    (let [running?       (atom true)
          now            (System/currentTimeMillis)
          lifespan       (+ now (rand-nth (range MIN-LIFESPAN MAX-LIFESPAN)))
          last-heartbeat (atom now)
          socket         (zmq/socket :dealer {:connect (str "tcp://*:" port)})]

      (zmq/after-shutdown
        (log/info "Shutting down."))

      (log/info "Sending READY signal.")
      (zmq/send-msg socket "READY")

      (zmq/polling {:stringify false}
        [socket :pollin [msg]
         (do
           (cond
           ; the server sends 1-frame messages as signals
            (= 1 (count msg))
            (let [signal (-> msg first (String.))]
              (case signal
                "KILL"      (do
                              (log/info "Received KILL signal from server.")
                              (reset! running? false))
                "STOP"      (do
                              (log/info "Received STOP signal from server.")
                              (stop-playback!))
                "HEARTBEAT" (do
                              (log/debug "Got HEARTBEAT from server.")
                              (reset! lives MAX-LIVES))
                (log/errorf "Invalid signal: %s" signal)))

           ; the server also forwards 3-frame messages from the client
           ; Frames:
           ;   1) the return address of the client
           ;   2) a JSON string representing the client's request
           ;   3) the command as a string (for use by the server)
            (= 3 (count msg))
            (let [[return-address body command] msg
                  body    (String. body)
                  command (String. command)]
              (try
                (when (and (not (available?))
                           (not (#{"play-status" "job-status"} command)))
                  (log/debugf "Rejecting message (command: %s). I'm busy." command)
                  (throw (Exception. "The requested worker is not available.")))
                (log/debugf "Processing message... (command: %s)" command)
                (let [req (json/parse-string body true)
                      res (json/generate-string (process req))]
                  (log/debug "Sending response...")
                  (zmq/send-msg socket [return-address "" res])
                  (log/debug "Response sent."))
                (catch Throwable e
                  (log/error e e)
                  (log/info "Sending error response...")
                  (let [err (json/generate-string (error-response e))]
                    (zmq/send-msg socket [return-address "" err]))
                  (log/info "Error response sent."))))

            :else
            (log/errorf "Invalid message: %s" (mapv #(String. %) msg))))]

        (while (and (zmq/polling?) @running?)
          (let [now      (System/currentTimeMillis)
                got-msgs (zmq/poll HEARTBEAT-INTERVAL)]
            (cond
              ;; Each worker has a randomly assigned lifespan in the range of
              ;; MIN-LIFESPAN and MAX-LIFESPAN. Once this period of time has
              ;; elapsed, the worker finishes whatever work it might be doing
              ;; and then shuts down so that the server can replace it with a
              ;; fresh worker.
              ;;
              ;; This ensures that the workers available are always recently
              ;; spawned processes, which helps us avoid known audio bugs.
              (and (> now lifespan) (available?))
              (do
                (log/info "Worker lifespan elapsed. Shutting down...")
                (reset! last-heartbeat now)
                (reset! running? false))

              ;; Detect when the system has been suspended and stop working so
              ;; the server can replace it with a fresh worker.
              ;;
              ;; This fixes a bug where MIDI audio is delayed.
              (> now (+ @last-heartbeat SUSPENDED-INTERVAL))
              (do
                (log/info "Process suspension detected. Shutting down...")
                (reset! last-heartbeat now)
                (reset! running? false))

              ;; If a heartbeat wasn't received from the server within the
              ;; acceptable threshold, MAX-LIVES times in a row, conclude that
              ;; the server has stopped sending heartbeats and shut down.
              (not (contains? got-msgs 0))
              (do
                (swap! lives dec)
                (when (and (<= @lives 0) (available?))
                  (log/error "Unable to reach the server.")
                  (reset! running? false))))

            ;; Send AVAILABLE/BUSY status back to the server.
            (when (and @running?
                       (> now (+ @last-heartbeat HEARTBEAT-INTERVAL)))
              (zmq/send-msg socket (if (available?) "AVAILABLE" "BUSY"))
              (reset! last-heartbeat now)))))))
  (exit! 0))
